docs/RFCS/20200331_enums.md

```diff
@@ -0,0 +1,298 @@
+
+This sort order byte array will be the physical layout and identifier of the enum. We expect
+that for small enums only a byte or two will be used to hold all the values, and that our
+compression strategies at the storage layer will compress this data well.
+
+An additional strategy is to represent enum values as a tuple of an integer and 
+this byte array.
+The ordering of the enums will be first on the integer, then on the byte array. An advantage
+of this approach is that adding a new enum to the beginning or end of the set requires only
+an increment/decrement of the integer rather than creating a new and possibly longer byte array.
+Adding to the beginning or end of the existing enum value set is the most
+common operation, and is the "worst case" for the pure byte array algorithm.
+A drawback of this strategy is that more space is used in the case where there are not many enum
+values. 
+
+## Parsing Changes
+
+Currently, the CockroachDB grammar is not equipped to handle type names
+that are qualified due to changes made in the past that separated parsing of
+object and type identifiers. Some of these changes will have to be 
+reverted/adapted in order to allow for types to have qualifications again.
+I am currently investigating the extend of the changes that need to be made in
+this area. An example of these changes are in 
+[this PR](https://github.com/cockroachdb/cockroach/pull/47216).
+
+
+## Type System Changes
+
+To update the type system to handle enums, we will introduce a new enum type
+which holds onto the ID of the enum. Using this, only enums of the same type
+can be compared etc. A new `Datum` `DEnum` will be introduced that holds 
+onto the ID of the enum that it represents as well as the physical layout 
+discussed above. Operations on two `DEnum` datums will succeed only if the
+two `DEnum`s have the same enum ID.
+
+The type system itself will also have to be extended to have support for
+resolved type references before type checking. The details involved in this
+change can be found in [this PR](https://github.com/cockroachdb/cockroach/pull/47386).
+
+When a user-defined type is created in Postgres, Postgres will automatically
+create an alias for an array of the new type. For example, if a user creates
+a type `days`, the system would also create the type `_days` as an alias for
+`days[]`. This type tracks changes made to the referenced type as it 
+moves through schemas and is dropped. It is unclear if supporting this sort
+of type aliasing will be a significant lift from the enum work.
+
+Each new enum type will need a unique OID. In order to guarantee that all
+user defined types have a unique OID we will maintain a new k/v level counter
+that is incremented transactionally as new types are created. This counter will
+start at a value larger than all existing Postgres type OIDs to avoid 
+collisions with existing types.
+
+## Optimizer Changes
```

Comment:
While `optbuilder` is owned by the optimizer team, it really has little to do with optimization. This should be called "Semantic Analysis Changes".